File: dk.brics.lightrefactor.experiments/src/dk/brics/lightrefactor/experiments/Equivalence.scala
package dk.brics.lightrefactor.experiments

import scala.collection.mutable

final class EquivNode[T](val item:T) {
  var parent = this
  var rank = 0
  
  def rep : EquivNode[T] = {
    if (parent == this) {
      this
    } else {
      val z = parent.rep
      parent = z
      z
    }
  }
}

/**
 * Computes an equivalence relation over T.
 */
class Equivalence[T] {
  val item2node = new mutable.HashMap[T,EquivNode[T]]
  private def getNode(x:T) = item2node.getOrElseUpdate(x, new EquivNode(x))
  def unify(x:T, y:T) {
    val xr = getNode(x).rep
    val yr = getNode(y).rep
    if (xr != yr) {
      val (high,low) = if (xr.rank > yr.rank) (xr,yr) else (yr,xr)
      low.parent = high
      if (high.rank == low.rank) {
        high.rank += 1
      }
    }
  }
  
  def getResult() : Map[T,T] = {
    var r = Map.empty[T,T]
    for ((t,node) <- item2node) {
      r += t -> node.rep.item
    }
    r
  }
}

File: dk.brics.lightrefactor/src/dk/brics/lightrefactor/types/Equivalence.java
package dk.brics.lightrefactor.types;

import java.util.HashMap;
import java.util.Map;

final class EquivalenceNode<T> {
  EquivalenceNode<T> parent = this;
  int rank = 0;
  T item;
  
  public EquivalenceNode(T item) {
    this.item = item;
  }
  
  public EquivalenceNode<T> rep() {
    if (parent == this) {
      return this;
    } else{
      return this.parent = parent.rep();
    }
  }
}

public class Equivalence<T> {
  
  private Map<T, EquivalenceNode<T>> map = new HashMap<T, EquivalenceNode<T>>();
  
  private EquivalenceNode<T> getNode(T x) {
    EquivalenceNode<T> node = map.get(x);
    if (node == null) {
      node = new EquivalenceNode<T>(x);
      map.put(x, node);
    }
    return node;
  }
  
  public boolean unify(T x, T y) {
    EquivalenceNode<T> xn = getNode(x).rep();
    EquivalenceNode<T> yn = getNode(y).rep();
    if (xn == yn)
      return false;
    if (xn.rank < yn.rank) {
      EquivalenceNode<T> z = xn;
      xn = yn;
      yn = z;
    } else if (xn.rank == yn.rank) {
      xn.rank += 1;
    }
    yn.parent = xn;
    return true;
  }
  
  public T getRep(T x) {
    return getNode(x).rep().item;
  }
  
  public Map<T,T> getResult() {
    Map<T,T> result = new HashMap<T,T>();
    for (Map.Entry<T, EquivalenceNode<T>> en : map.entrySet()) {
      result.put(en.getKey(), en.getValue().rep().item);
    }
    return result;
  }
}
